specs/LaunchingRegisteredProtocols.md

```diff
@@ -0,0 +1,264 @@
+[uuid(fc43b557-9713-4a67-af8d-a76ef3a206e8), object, pointer_default(unique)] 
+interface ICoreWebView2LaunchingRegisteredProtocolEventArgs: IUnknown {
+  /// The uri of the requested protocol.
+
+  [propget] HRESULT Uri([out, retval] LPWSTR* uri);
+
+  /// The host may set this flag to cancel the protocol launch.  If set to
+  /// `TRUE`, the protocol will not be launched. If cancelled, the
+  /// dialog is not displayed regardless of the `Handled` property.
+
+  [propget] HRESULT Cancel([out, retval] BOOL* cancel);
+
+  /// Sets the `Cancel` property.
+
+  [propput] HRESULT Cancel([in] BOOL cancel);
+
+  /// The host may set this flag to `TRUE` to hide the dialog.
+  /// The protocol request will continue as normal if it is not
+  /// cancelled, although there will be no default UI shown. By default the
+  /// value is `FALSE` and the default registered protocol dialog is shown.
+
+  [propget] HRESULT Handled([out, retval] BOOL* handled);
+
+  /// Sets the `Handled` property.
+
+  [propput] HRESULT Handled([in] BOOL handled);
+
+  /// The host may set this flag to `TRUE` to revoke previous permissions given
+  /// to this protocol/origin. In the case in which the registered protocol
+  /// launch is being requested from a trustworthy origin, the dialog
+  /// prompt (if not handled) will display a checkbox that gives the user the option to
+  /// always allow this origin to open links of this type in the associated app.
+  /// Once checked by the user, this protocol for this origin will be launched automatically
+  /// without displaying the dialog prompt.
+  /// If the `Cancel` property is set to `TRUE` while this property is set to
+  /// `TRUE` the registered protocol will not be launched in this instance. If `Handled` is set  
+  /// to `TRUE` the dialog prompt will still be disabled regardless of this property, but the 
+  /// permissions will be revoked in the future if this property is set to `TRUE`. 
+
+  [propget] HRESULT RevokeProtocolPermissionsPerOrigin(
+      [out, retval] BOOL* revokeProtocolPermissionsPerOrigin);
+
+  /// Sets the `RevokeProtocolPermissionsPerOrigin` property.
+  [propput] HRESULT RevokeProtocolPermissionsPerOrigin([in] BOOL revokeProtocolPermissionsPerOrigin);
+}
+
+``` 
+## .NET and WinRT
+
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    runtimeclass CoreWebView2LaunchingRegisteredProtocolEventArgs;
+    
+    runtimeclass CoreWebView2LaunchingRegisteredProtocolEventArgs
+    {
+        // CoreWebView2LaunchingRegisteredProtocolEventArgs members
+        String Uri { get; };
+        Boolean Cancel { get; set; };
+        Boolean Handled {get; set; };
+        Boolean RevokeProtocolPermissionsPerOrigin {get; set; };
+    }
+
+    runtimeclass CoreWebView2
+    {
+        // CoreWebView2 
+        event Windows.Foundation.TypedEventHandler<CoreWebView2, CoreWebView2LaunchingRegisteredProtocolEventArgs> LaunchingRegisteredProtocolEventArgs;
+    }
+
+    runtimeclass CoreWebView2Frame
```

Comment:
We currently only have a way to get top level frames, not frames within frames. Is this a problem for your scenario? As in, the end dev won't be able to intercept these protocol launches for frames within frames. Is the frame context important, or can we just have the one CoreWebView2.LaunchingRegisteredProtocol event raised no matter which frame was trying to do the navigation?

Reply:
This isn't a problem for my scenario.  Office add in would like to know whether or not the request is originating from an iframe or not. Do you mean removing the frame event entirely and only having the event fire from CoreWebView2?

Reply:
I wasn't suggesting a change necessarily, just pointing out that currently you can only get a CoreWebView2Frame for an iframe in the top level document and you can't get a CoreWebView2Frame for an iframe inside an iframe. If an iframe in an iframe navigates to `calculator:` or whatever, what will happen with your API?

Reply:
The event will fire from the CoreWebView2 interface as well as from the top level frame. The first interface to handle the event will determine the protocol launch behavior.

Reply:
If a nested iframe navigates to calculator:, the event will be fired from the top most iframe.